setup.py

```diff
@@ -37,14 +37,14 @@ def find_name() -> str:
     raise RuntimeError("Unable to find name string.")
 
 
-def find_long_description():
+def find_long_description() -> str:
     """Return the content of the README.md file."""
     return read_file("../README.md")
 
 
 # upper limits are untested, not necessarily conflicting
 # lower limits mostly to be python 3 compatible
-REQUIRED_PACKAGES = ["numpy>=0.16.0,<2.0.0", "pygtrie>=2.1,<3.0", "hypothesis>=6.14,<7"]
+REQUIRED_PACKAGES = ["numpy>=0.21.0,<2.0.0", "pygtrie>=2.1,<3.0", "hypothesis>=6.14,<7"]
```

Comment:
what exactly is needed here? this is a muuch tighter restriction just for some mypy annotations. any way to get around it?

Reply:
Yeah thoughts welcome here.  Firstly that's a typo-- it should be 1.20.  1.0 was released in 2006 and I'd be really surprised if anyone were still using 0.x.

1.20 (released in January 2021) introduced `np.typing` which (AIUI) is supposed to put us on a path towards better type support for ndarrays, including checked dimensions.   Annotating with np.ndarray becomes an error: one must use the parametrized `NDArray` instead.  On the other hand, if the user has numpy < 1.20 installed, the typing module will be missing.

Options:
- bumping the lower bound to 1.20
- pinning the upper bound at 1.19
- We could try introspection hackery to define an appropriate type alias depending on version.  This seems tough to do statically.

In any case, the existing bound of 0.16 is way too low.  0.16 is not even pip-installable anymore :/